core/src/main/scala/caliban/wrappers/CostEstimation.scala

```diff
@@ -0,0 +1,257 @@
+   * @param f The function used to evaluate the cost of a single field
+   * @param error A function that will be provided the total estimated cost and must return the error that will be returned
+   */
+  def maxCostOrError(maxCost: Double)(f: Field => Double)(error: Double => ValidationError): ValidationWrapper[Any] =
+    new ValidationWrapper[Any] {
+      override def wrap[R1 <: Any](
+        process: Document => ZIO[R1, ValidationError, ExecutionRequest]
+      ): Document => ZIO[R1, ValidationError, ExecutionRequest] =
+        (doc: Document) =>
+          for {
+            req <- process(doc)
+            cost = computeCost(req.field)(f)
+            _   <- IO.when(cost > maxCost)(IO.fail(error(cost)))
+          } yield req
+    }
+
+  /**
+   * More powerful version of [[maxCost]] which allows the field computation function to specify a function which returns an effectful computation
+   * for the cost of a field.
+   * @param maxCost The total cost allowed for any one query
+   * @param f The function used to evaluate the cost of a single field returning an effect which will result in the field cost as a double
+   */
+  def maxCostZIO[R](maxCost: Double)(f: Field => URIO[R, Double]): ValidationWrapper[R] =
+    new ValidationWrapper[R] {
+      override def wrap[R1 <: R](
+        process: Document => ZIO[R1, ValidationError, ExecutionRequest]
+      ): Document => ZIO[R1, ValidationError, ExecutionRequest] =
+        (doc: Document) =>
+          for {
+            req  <- process(doc)
+            cost <- computeCostZIO(req.field)(f)
+            _    <- IO.when(cost > maxCost)(
+                      IO.fail(ValidationError(s"Query costs too much: $cost. Max cost: $maxCost.", ""))
+                    )
+          } yield req
+    }
+
+  private def costWrapper(total: Ref[Double])(f: Field => Double): ValidationWrapper[Any] =
+    new ValidationWrapper[Any] {
+      override def wrap[R1 <: Any](
+        process: Document => ZIO[R1, ValidationError, ExecutionRequest]
+      ): Document => ZIO[R1, ValidationError, ExecutionRequest] =
+        (doc: Document) =>
+          for {
+            req <- process(doc)
+            _   <- total.set(computeCost(req.field)(f))
+          } yield req
+    }
+
+  private def costWrapperZIO[R](total: Ref[Double])(f: Field => URIO[R, Double]): ValidationWrapper[R] =
+    new ValidationWrapper[R] {
+      override def wrap[R1 <: R](
+        process: Document => ZIO[R1, ValidationError, ExecutionRequest]
+      ): Document => ZIO[R1, ValidationError, ExecutionRequest] =
+        (doc: Document) =>
+          for {
+            req <- process(doc)
+            _   <- computeCostZIO(req.field)(f) >>= total.set
+          } yield req
+    }
+
+  private def addCostToExtensions(
+    cost: Ref[Double]
+  )(resp: GraphQLResponse[CalibanError]): UIO[GraphQLResponse[CalibanError]] =
+    cost.get.map { total =>
+      resp.copy(
+        extensions = Some(
+          ObjectValue(
```

Comment:
This is a really nice pattern! Previously I've used HTTP headers to communicate things like rate limits but this is much nicer.